### PR TITLE
perf: loops optimization

### PR DESCRIPTION
**Pull request by simone-sanfratello**
npm run bench this branch

[1] http - no agent  x 711 ops/sec ±0.86% (84 runs sampled)
[1] http - keepalive x 760 ops/sec ±0.64% (78 runs sampled)
[1] http - keepalive - multiple sockets x 4,732 ops/sec ±1.70% (84 runs sampled)
[1] undici - pipeline x 5,531 ops/sec ±1.81% (84 runs sampled)
[1] undici - request x 5,656 ops/sec ±0.77% (83 runs sampled)
[1] undici - pool - request - multiple sockets x 5,720 ops/sec ±0.64% (86 runs sampled)
[1] undici - stream x 5,809 ops/sec ±0.60% (84 runs sampled)
[1] undici - dispatch x 5,986 ops/sec ±0.45% (88 runs sampled)
[1] undici - noop x 7,414 ops/sec ±0.34% (87 runs sampled)

npm run bench main branch

[1] http - no agent  x 719 ops/sec ±0.73% (84 runs sampled)
[1] http - keepalive x 747 ops/sec ±0.52% (79 runs sampled)
[1] http - keepalive - multiple sockets x 4,644 ops/sec ±1.91% (81 runs sampled)
[1] undici - pipeline x 5,437 ops/sec ±2.06% (80 runs sampled)
[1] undici - request x 5,624 ops/sec ±1.01% (84 runs sampled)
[1] undici - pool - request - multiple sockets x 5,639 ops/sec ±0.59% (85 runs sampled)
[1] undici - stream x 5,856 ops/sec ±0.72% (85 runs sampled)
[1] undici - dispatch x 6,046 ops/sec ±0.58% (86 runs sampled)
[1] undici - noop x 7,342 ops/sec ±0.28% (89 runs sampled)

### CODE REVIEW

File: lib/agent/redirect.js [hunk 1]
```diff
@@ -143,15 +143,17 @@ function shouldRemoveHeader (header, removeContent) {
...
+    const length = headers.length
+    for (let i = 0; i < length; i += 2) {
+      const header = headers[i]
+      if (!shouldRemoveHeader(header, removeContent)) {
+        ret.push(header, headers[i + 1])
       }
     }
   } else if (headers && typeof headers === 'object') {
-    for (const [key, val] of Object.entries(headers)) {
+    for (const key in headers) {
```
Comment (mcollina) [inline lib/agent/redirect.js:154]: please revert this change. Using `in` makes it vulnerable to prototype pollution attacks.

File: lib/client-pool.js [hunk 2]
```diff
@@ -147,8 +153,10 @@ class Pool extends EventEmitter {
   get size () {
     let ret = this[kPending]
 
-    for (const { size } of this[kClients]) {
-      ret += size
+    const clients = this[kClients]
+    const length = clients.length
+    for (let i = 0; i < length; i++) {
```
Comment (mcollina) [inline lib/client-pool.js:158]: please use `for (var`, it's faster.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: lib/agent/redirect.js (hunk 1 above)
Comment (szmarczak) [reply lib/agent/redirect.js:154]: But the object is given by the user. I think it should be their responsibility to prevent prototype pollution.

Comment (szmarczak) [reply lib/agent/redirect.js:154]: Alternatively `hasOwnProperty` could be used. Or `Object.keys` instead of `Object.entries`.

Comment (mcollina) [reply lib/agent/redirect.js:154]: Object.keys is the better option?

File: lib/client.js [hunk 3]
```diff
@@ -547,14 +550,17 @@ class Parser extends HTTPParser {
     const { headers } = this
     this.headers = null
 
-    for (let n = 0; n < headers.length; n += 2) {
-      const key = headers[n + 0]
+    let looking = true
+    for (let n = 0; n < headers.length && looking; n += 2) {
```
Comment (mcollina) [inline lib/client.js:554]: for (var

File: lib/client.js [hunk 4]
```diff
@@ -695,10 +701,13 @@ class Parser extends HTTPParser {
           throw new TrailerMismatchError()
         }
 
-        for (const trailer of trailers) {
+        const trailersLength = trailers.length
+        for (let i = 0; i < trailersLength; i++) {
```
Comment (mcollina) [inline lib/client.js:705]: for (var

File: examples/proxy/proxy.js [hunk 5]
```diff
@@ -134,7 +134,7 @@ class WSHandler {
 
     let head = ''
     for (let n = 0; n < headers.length; n += 2) {
-      head += `\r\n${headers[n + 0]}: ${headers[n + 1]}`
+      head += `\r\n${headers[n]}: ${headers[n + 1]}`
```
Comment (mcollina) [inline examples/proxy/proxy.js:137]: there is a for (var to be fixed above

**Review by mcollina — COMMENTED** [review]
Can you check if this is giving a warning from `standard`?

**Comment by simone-sanfratello** [comment]
> Can you check if this is giving a warning from `standard`?

yes, it is. Should I use eslint and extend standard, then allow vars?

**Comment by ronag** [comment]
Is var actually faster? Do we have any numbers?

**Comment by simone-sanfratello** [comment]
> Can you check if this is giving a warning from `standard`?

with standard `v.14`, used in the project, is fine, but with the `v.16`, the latest, is not

File: lib/client.js [hunk 6]
```diff
@@ -349,8 +349,10 @@ class Client extends Dispatcher {
       err = new ClientDestroyedError()
     }
 
-    for (const request of this[kQueue].splice(this[kPendingIdx])) {
-      request.onError(err)
+    const requests = this[kQueue].splice(this[kPendingIdx])
+    const length = requests.length
+    for (var i = 0; i < length; i++) {
```
Comment (tbnritzdoge) [inline lib/client.js:354]:
```suggestion
    for (var i = 0; i < length; ++i) {
```
faster

File: lib/client.js [hunk 7]
```diff
@@ -360,8 +362,9 @@ class Client extends Dispatcher {
     const onDestroyed = () => {
       const callbacks = this[kOnDestroyed]
       this[kOnDestroyed] = null
-      for (const callback of callbacks) {
-        callback(null, null)
+      const length = callbacks.length
+      for (var i = 0; i < length; i++) {
```
Comment (tbnritzdoge) [inline lib/client.js:366]:
```suggestion
      for (var i = 0; i < length; ++i) {
```
faster

File: lib/client.js [hunk 8]
```diff
@@ -695,10 +701,13 @@ class Parser extends HTTPParser {
           throw new TrailerMismatchError()
         }
 
-        for (const trailer of trailers) {
+        const trailersLength = trailers.length
+        for (var i = 0; i < trailersLength; i++) {
```
Comment (tbnritzdoge) [inline lib/client.js:705]:
```suggestion
        for (var i = 0; i < trailersLength; ++i) {
```
faster

File: lib/client.js [hunk 9]
```diff
@@ -835,8 +846,10 @@ function onSocketClose () {
   if (client[kDestroyed]) {
     assert(client.pending === 0)
 
-    for (const request of client[kQueue].splice(client[kRunningIdx])) {
-      request.onError(err)
+    const requests = client[kQueue].splice(client[kRunningIdx])
+    const length = requests.length
+    for (var i = 0; i < length; i++) {
```
Comment (tbnritzdoge) [inline lib/client.js:851]:
```suggestion
    for (var i = 0; i < length; ++i) {
```

File: lib/core/request.js [hunk 10]
```diff
@@ -66,12 +66,16 @@ class Request {
...
+      const length = headers.length
+      for (var i = 0; i < length; i += 2) {
+        processHeader(this, headers[i], headers[i + 1])
       }
     } else if (headers && typeof headers === 'object') {
-      for (const [key, val] of Object.entries(headers)) {
-        processHeader(this, key, val)
+      const keys = Object.keys(headers)
+      const length = keys.length
+      for (var j = 0; j < length; j++) {
```
Comment (tbnritzdoge) [inline lib/core/request.js:76]:
```suggestion
      for (var j = 0; j < length; ++j) {
```
faster

File: lib/llhttp/parser.js [hunk 11]
```diff
@@ -171,7 +171,8 @@ class HTTPParser {
 
   [kGetHeaders] () {
     const rawHeaders = []
-    for (let c = 0; c < this[kHeadersFields].length; c++) {
+    const length = this[kHeadersFields].length
+    for (var c = 0; c < length; c++) {
```
Comment (tbnritzdoge) [inline lib/llhttp/parser.js:175]:
```suggestion
    for (var c = 0; c < length; ++c) {
```
faster

File: lib/pool.js [hunk 12]
```diff
@@ -126,8 +126,10 @@ class Pool extends Dispatcher {
   get pending () {
     let ret = this[kPending]
 
-    for (const { pending } of this[kClients]) {
-      ret += pending
+    const clients = this[kClients]
+    const length = clients.length
+    for (var i = 0; i < length; i++) {
```
Comment (tbnritzdoge) [inline lib/pool.js:131]:
```suggestion
    for (var i = 0; i < length; ++i) {
```

File: lib/pool.js [hunk 13]
```diff
@@ -136,8 +138,10 @@ class Pool extends Dispatcher {
   get running () {
     let ret = 0
 
-    for (const { running } of this[kClients]) {
-      ret += running
+    const clients = this[kClients]
+    const length = clients.length
+    for (var i = 0; i < length; i++) {
```
Comment (tbnritzdoge) [inline lib/pool.js:143]:
```suggestion
    for (var i = 0; i < length; ++i) {
```

File: lib/pool.js [hunk 14]
```diff
@@ -146,8 +150,10 @@ class Pool extends Dispatcher {
   get size () {
     let ret = this[kPending]
 
-    for (const { size } of this[kClients]) {
-      ret += size
+    const clients = this[kClients]
+    const length = clients.length
+    for (var i = 0; i < length; i++) {
```
Comment (tbnritzdoge) [inline lib/pool.js:155]:
```suggestion
    for (var i = 0; i < length; ++i) {
```

**Review by tbnritzdoge — CHANGES_REQUESTED** [review]
Replace all i++ 's with ++i as it is faster and more memory efficient (doesn't make a copy)

**Comment by szmarczak** [comment]
> Replace all i++ 's with ++i as it is faster and more memory efficient (doesn't make a copy)

@tbnritzdoge I'd have never thought of that 🤯 Nice spot.

**Comment by ronag** [comment]
> Replace all i++ 's with ++i as it is faster and more memory efficient (doesn't make a copy)

That’s somewhat true with c++ iterators but doesn’t really apple here.

**Comment by szmarczak** [comment]
> That’s somewhat true with c++ iterators but doesn’t really apple here.

I guess V8 optimizes this?

**Comment by Nytelife26** [comment]
```
prefix x 522,745,219 ops/sec ±1.32% (85 runs sampled)
suffix x 516,346,780 ops/sec ±1.19% (84 runs sampled)
```
Having ran a benchmark between the two, `i++` is slightly slower usually. 

However, having ran another benchmark,

> please use `for (var`, it's faster.

```
for-of x 68,941,007 ops/sec ±1.07% (87 runs sampled)
c-style (prefix) x 92,530,979 ops/sec ±0.57% (96 runs sampled)
c-style (suffix) x 92,918,425 ops/sec ±0.45% (97 runs sampled)
```

for-of turns out to be significantly slower. Aside from that, it is largely irrelevant which is used, be it prefix or suffix, because the performance difference is inconsistent and negligible. It may be prevalent to use the prefix method though, given that, as Doge rightly pointed out, it creates an extra copy on the stack.

**Comment by simone-sanfratello** [comment]
> for-of turns out to be significantly slower. Aside from that, it is largely irrelevant which is used, be it prefix or suffix, because the performance difference is inconsistent and negligible. It may be prevalent to use the prefix method though, given that, as Doge rightly pointed out, suffix creates an extra copy on the stack which leads to unnecessary memory usage.

I'm doing some (a lot) benchmarks too, yes for-of is quite slower than "classic" for
I'll add the `++i` to the cases, but looks like it doesn't make much difference

**Comment by wzhouwzhou** [comment]
Agreed with @ronag, as usage of ++i and i++ here is likely optimised away with return value dropped, and any turbofan v8 loop [unrolling](https://github.com/v8/v8/blob/dc712da548c7fb433caed56af9a021d964952728/src/codegen/code-stub-assembler.h#L3706) will also condense that down into at most a couple of instructions. However, those optimisations may only occur in some picture-perfect cases. 

At this point, we should always consider whether we are venturing into premature optimisation territory, as with web-oriented applications and libraries, the bottlenecks will almost certainly occur elsewhere. I would hate for a library to become any less maintainable or inconsistent with itself (i.e., i++'s here, ++i's there, lack of structure and order) over changes like that. (Same issue as the fastify reference above).

**Comment by tbnritzdoge** [comment]
- [ ] You should note that i++ and ++i are  actually not the same thing and are different things hence force the different optimization levels. In addition to your point I think anyone with a interest in these libraries has a obvious wish to reach the pinnacle of performance and these incremental changes push us to that edge, but if your so insistent on your techniques you should consider using slower and "cleaner" alternatives such as the built in http(s) and express libraries @wzhouwzhou

**Comment by wzhouwzhou** [comment]
@tbnritzdoge We're really gratefulfor your interest in making this library faster! As mentioned previously, we see contradictory test results with no obvious improvements. That being said, for those wishing to reach the actual "pinnacle of performance" I recommend you learn languages such as C++ and rust! One of the most important but overlooked aspects of library maintenance and open source design involves maintainability, a skill that takes time and real-world exposure for you to experience. 
Looking forward to see your growth in the future!

**Comment by tbnritzdoge** [comment]
> @tbnritzdoge We're really grateful for your interest in making this library faster! As mentioned previously, we see contradictory test results with no obvious improvements. That being said, for those wishing to reach the actual "pinnacle of performance" I recommend you learn languages such as C++ and rust! One of the most important but overlooked aspects of library maintenance and open source design involves maintainability, a skill that takes time and real-world exposure for you to experience. 
> 
> Looking forward to see your growth in the future!

1. I don't see any memory benchmarks nor do I see a understanding of node's internals as its very obvious that having a non copying function is better than a copying one.

2. I do indeed spend a lot of my time writing rust, but if you worked for a real company you would know that porting an entire codebase is a immense chore so it's better in the short term to find simple optimizations on what you have rather than scrap everything and rewrite to a new language.

**Comment by wzhouwzhou** [comment]
1. You should take a look into how ignition and turbofan really work - I recommend you start with some v8 blogs available on google. Given any impressive rust background, the conversion from bytecode to machine code should be easy for you.
2. The overhead of porting codebases is irrelevant to the conversation - if you worked at a real company you would know that we are talking about a series of tradeoffs and performance on orders of magnitudes worth of difference, and that the juxtaposition is the critical focus, rather than the execution. Both enterprise level code and typical users of NodeJS use scripting language for ease of use. It's better in the short-term for you to focus on areas of the library that actually need improvement to design and control flow with known bottlenecks rather than make a big deal out of a subjective convention with insufficient proof.

**Comment by tbnritzdoge** [comment]
> 1. You should take a look into how ignition and turbofan really work - I recommend you start with some v8 blogs available on google. Given any impressive rust background, the conversion from bytecode to machine code should be easy for you.
> 
> 2. The overhead of porting codebases is irrelevant to the conversation - if you worked at a real company you would know that we are talking about a series of tradeoffs and performance on orders of magnitudes worth of difference, and that the juxtaposition is the critical focus, rather than the execution. Both enterprise level code and typical users of NodeJS use scripting language for ease of use. It's better in the short-term for you to focus on areas of the library that actually need improvement to design and control flow with known bottlenecks rather than make a big deal out of a subjective convention with insufficient proof.

If you actually were as proficient as you act you are you would know the performance benefits.

**Comment by wzhouwzhou** [comment]
At this point it seems that you've entirely strayed away from any argument grounded in theory and reason - understanding proficiency as well as the multiple facets of performance myself (while taking into account developer efficiency), it looks like there is nothing more to be discussed with you in your current state. Please go touch some grass and consider what gives you that right.

For what it's worth, the decision on this PR has already been made already, so it's confusing why you belabour your suggestions and express significant amounts of discontent at any discussions regarding them. It's childish and unprofessional. We are just having a light chat about the pros and cons.

Kindly don't make an immense chore out of this and any future contributions to NodeJS in the future. Thanks!

**Comment by tbnritzdoge** [comment]
> At this point it seems that you've entirely strayed away from any argument grounded in theory and reason - understanding proficiency as well as the multiple facets of performance myself (while taking into account developer efficiency), it looks like there is nothing more to be discussed with you in your current state. Please go touch some grass and consider what gives you that right.
> 
> 
> 
> For what it's worth, the decision on this PR has already been made already, so it's confusing why you belabour your suggestions and express significant amounts of discontent at any discussions regarding them. It's childish and unprofessional. We are just having a light chat about the pros and cons.
> 
> 
> 
> Kindly don't make an immense chore out of this and any future contributions to NodeJS in the future. Thanks!

I will certainly go outside and touch some grass, thanks for the suggestion. I hope you have learned something about performance critical nodejs code and that you're able to use this knowledge in your own code. :)

**Comment by tbnritzdoge** [comment]
> At this point it seems that you've entirely strayed away from any argument grounded in theory and reason - understanding proficiency as well as the multiple facets of performance myself (while taking into account developer efficiency), it looks like there is nothing more to be discussed with you in your current state. Please go touch some grass and consider what gives you that right.
> 
> 
> 
> For what it's worth, the decision on this PR has already been made already, so it's confusing why you belabour your suggestions and express significant amounts of discontent at any discussions regarding them. It's childish and unprofessional. We are just having a light chat about the pros and cons.
> 
> 
> 
> Kindly don't make an immense chore out of this and any future contributions to NodeJS in the future. Thanks!

I will certainly go outside and touch some grass, thanks for the suggestion. I hope you have learned something about performance critical nodejs code and that you're able to use this knowledge in your own code. :)

**Comment by wzhouwzhou** [comment]
Glad to hear, you have a good day now mate. I'll be discussing the pros and cons in my next book, tentatively titled Learning to Code the Wrong Way, so definitely hope you'll check it out 👍

**Comment by tbnritzdoge** [comment]
> Glad to hear, you have a good day now mate. I'll be discussing the pros and cons in my next book, tentatively titled Learning to Code the Wrong Way, so definitely hope you'll check it out 👍  

Will be looking forward to your book, gladly you're writing about a subject you know a lot about in contradiction to many authors in today's society. :)

**Comment by wzhouwzhou** [comment]
Glad to have you on board, I'll let you know when it's finished. We'll make wrong the new right 👍

**Comment by simone-sanfratello** [comment]
Please note the purpose of this PR is to optimize loops, without going too far.

I'm doing a lot of benchmarks, at the moment I see there is not a relevant difference between using `let i` or `var i` in for loops, while `var` is slightly better (especially with small arrays as these ones), I would prefer `let` to avoid tweaking the linter. @mcollina 

`Object.keys` is a way better than  Object.values` (40% faster), no doubt about that.

I'll publish benchmark results as soon as I can.

**Comment by mcollina** [comment]
> Please note the purpose of this PR is to optimize loops, without going too far.
> 
> 
> 
> I'm doing a lot of benchmarks, at the moment I see **there is no a relevant difference** between using `let i` or `var i` in for loops, while `var` is slightly better (especially with small arrays as these ones), I would prefer `let` to avoid tweaking the linter. @mcollina 

The last time I checked on Node v12 the difference was still there. Node v12 is still supported.

**Comment by wzhouwzhou** [comment]
Thanks so much @simone-sanfratello

**Comment by mcollina** [comment]
Let me correct my take as I found my old measurements: there are no differences between var and let in v12, so +1 to move it. More info at https://github.com/fastify/fastify/pull/2667#issuecomment-720592319.

**Comment by simone-sanfratello** [comment]
These the result of a week of benchmarks

https://github.com/simone-sanfratello/node-bench-iteration

happy to update the code to `let` also to remove my proposal to use `n = array.length ... for(.. i < n)`

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by simone-sanfratello** [comment]
I don't get where the coverage is dropping, can anyone help me?

**Comment by ronag** [comment]
CI fails. Maybe try to rebase main?

**Comment by simone-sanfratello** [comment]
it's already rebased to main

**Comment by ronag** [comment]
I'll try re-running the CI jobs

**Comment by ronag** [comment]
Was a flaky test. Thanks!
